src/shims/ffi_support.rs

```diff
@@ -0,0 +1,285 @@
+                    != _lib_path.to_str().unwrap()
+                {
+                    return None;
+                }
+            }
+        }
+        // Return a pointer to the function.
+        Some(CodePtr(*func.deref() as *mut _))
+    }
+
+    /// Call specified external C function, with supplied arguments.
+    /// Need to convert all the arguments from their hir representations to
+    /// a form compatible with C (through `libffi` call).
+    /// Then, convert return from the C call into a corresponding form that
+    /// can be stored in Miri internal memory.
+    fn call_and_add_external_c_fct_to_context(
+        &mut self,
+        link_name: Symbol,
+        dest: &PlaceTy<'tcx, Provenance>,
+        args: &[OpTy<'tcx, Provenance>],
+    ) -> InterpResult<'tcx, bool> {
+        // Get the pointer to the function in the shared object file if it exists.
+        let code_ptr = match self.get_func_ptr_explicitly_from_lib(link_name) {
+            Some(ptr) => ptr,
+            None => {
+                // Shared object file does not export this function -- try the shims next.
+                return Ok(false);
+            }
+        };
+
+        let this = self.eval_context_mut();
+
+        // Get the function arguments, and convert them to `libffi`-compatible form.
+        let mut libffi_args = Vec::<CArg>::with_capacity(args.len());
+        for cur_arg in args.iter() {
+            libffi_args.push(Self::scalar_to_carg(
+                this.read_scalar(cur_arg)?,
+                &cur_arg.layout.ty,
+                this,
+            )?);
+        }
+
+        // Convert them to `libffi::high::Arg` type.
+        let libffi_args = libffi_args
+            .iter()
+            .map(|cur_arg| cur_arg.arg_downcast())
+            .collect::<Vec<libffi::high::Arg<'_>>>();
+
+        // Code pointer to C function.
+        // let ptr = CodePtr(*func.deref() as *mut _);
+        // Call the function and store output, depending on return type in the function signature.
+        self.call_external_c_and_store_return(link_name, dest, code_ptr, libffi_args)?;
+        Ok(true)
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Enum of supported arguments to external C functions.
+pub enum CArg {
+    /// 8-bit signed integer.
+    Int8(i8),
+    /// 16-bit signed integer.
+    Int16(i16),
+    /// 32-bit signed integer.
+    Int32(i32),
+    /// 64-bit signed integer.
+    Int64(i64),
+    /// isize.
+    ISize(isize),
+    /// 8-bit unsigned integer.
+    UInt8(u8),
+    /// 16-bit unsigned integer.
+    UInt16(u16),
+    /// 32-bit unsigned integer.
+    UInt32(u32),
+    /// 64-bit unsigned integer.
+    UInt64(u64),
+    /// usize.
+    USize(usize),
+}
+
+impl<'a> CArg {
+    /// Convert a `CArg` to a `libffi` argument type.
+    pub fn arg_downcast(&'a self) -> libffi::high::Arg<'a> {
+        match self {
+            CArg::Int8(i) => arg(i),
```

Comment:
This is from libffi, right? Please import it in a way that this ends up being more like `ffi::arg` or so, making it clear that we are handing off to libffi here.